dmypy issue with `mypy.ini`

In http://github.com/zulip/zulip/, we've been running `dmypy` with this commit from master: https://github.com/python/mypy.git@6a934d32124991a77ea433856cf2d62871ceb632#egg=mypy==0.610+dev.6a934d32124991a77ea433856cf2d62871ceb632

Generally, it's been great, but we've been prevented from turning it on by default because I've been seeing it throw occasional errors like this (generally shows up after a lot of rebasing; I switch branches dozens of times a day as part of merging things):

```
(zulip-py3-venv) tabbott@zaset:~/zulip$ ./tools/run-mypy -d
zproject/test_settings.py:22: error: Need type annotation for 'REALM_HOSTS'
zproject/test_settings.py:43: error: Incompatible types in assignment (expression has type "None", variable has type "str")
zproject/test_settings.py:61: error: Incompatible types in assignment (expression has type "Tuple[str, ...]", variable has type "Tuple[str, str, str, str]")
zproject/test_settings.py:80: error: Dict entry 2 has incompatible type "str": "int"; expected "str": "Union[Dict[str, int], str, None]"
zproject/test_settings.py:80: error: Dict entry 3 has incompatible type "str": "int"; expected "str": "Union[Dict[str, int], str, None]"
zproject/test_settings.py:152: error: Need type annotation for 'PUSH_NOTIFICATION_BOUNCER_URL'
```

but this shouldn't be happening, since we have this in our `mypy.ini`:

```
[mypy-zproject.test_settings]                                                                                                            
ignore_errors = True                                                                                                                     
```

I have it reproduced currently in my laptop, so if there's something I can do to dump data from `dmypy` or otherwise help debug, I'd love to do that.

As a sidenote, why `dmypy` instead of `mypyd`?  I feel like the latter is the traditional UNIX name :)

--- a/mypy/semanal_pass3.py
+++ b/mypy/semanal_pass3.py
@@ -77,6 +77,7 @@ class SemanticAnalyzerPass3(TraverserVisitor, SemanticAnalyzerCoreInterface):
     def refresh_partial(self, node: Union[MypyFile, FuncItem, OverloadedFuncDef],
                         patches: List[Tuple[int, Callable[[], None]]]) -> None:
         """Refresh a stale target in fine-grained incremental mode."""
+        self.options = self.sem.options
         self.patches = patches
         if isinstance(node, MypyFile):
             self.recurse_into_functions = False

--- a/mypy/server/update.py
+++ b/mypy/server/update.py
@@ -852,9 +852,6 @@ def reprocess_nodes(manager: BuildManager,
 
     nodes = sorted(nodeset, key=key)
 
-    # TODO: ignore_all argument to set_file_ignored_lines
-    manager.errors.set_file_ignored_lines(file_node.path, file_node.ignored_lines)
-
     targets = set()
     for node in nodes:
         target = target_from_node(module_id, node.node)

--- a/mypy/server/update.py
+++ b/mypy/server/update.py
@@ -852,6 +852,10 @@ def reprocess_nodes(manager: BuildManager,
 
     nodes = sorted(nodeset, key=key)
 
+    options = graph[module_id].options
+    manager.errors.set_file_ignored_lines(
+        file_node.path, file_node.ignored_lines, options.ignore_errors)
+
     targets = set()
     for node in nodes:
         target = target_from_node(module_id, node.node)
@@ -868,7 +872,6 @@ def reprocess_nodes(manager: BuildManager,
 
     # Second pass of semantic analysis. We don't redo the first pass, because it only
     # does local things that won't go stale.
-    options = graph[module_id].options
     for deferred in nodes:
         with semantic_analyzer.file_context(
                 file_node=file_node,

--- a/test-data/unit/fine-grained.test
+++ b/test-data/unit/fine-grained.test
@@ -7139,3 +7139,55 @@ def f() -> int: return 0
 def f() -> str: return '0'
 [out]
 ==
+
+[case testRefreshIgnoreErrors1]
+[file mypy.ini]
+[[mypy]
+[[mypy-b]
+ignore_errors = True
+[file a.py]
+y = '1'
+[file a.py.2]
+y = 1
+[file b.py]
+from a import y
+def fu() -> None:
+    1+'lurr'
+    y
+[out]
+==
+
+[case testRefreshIgnoreErrors2]
+[file mypy.ini]
+[[mypy]
+[[mypy-b]
+ignore_errors = True
+[file b.py]
+def fu() -> int:
+    1+'lurr'
+    return 1
+[file b.py.2]
+def fu() -> int:
+    1+'lurr'
+    return 2
+[out]
+==
+
+
+[case testRefreshOptions]
+[file mypy.ini]
+[[mypy]
+disallow_any_generics = True
+[[mypy-b]
+disallow_any_generics = False
+[file a.py]
+y = '1'
+[file a.py.2]
+y = 1
+[file b.py]
+from typing import List
+from a import y
+x = []  # type: List
+[builtins fixtures/list.pyi]
+[out]
+==

